Implement undo/redo at cell level

barbaravaldez opened the pull request:
Implemented undo and redo for adding, deleting and moving cells. 
This PR fixes #15049 and fixes #8011

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/notebookHistory.ts:
```diff
@@ -0,0 +1,71 @@
+	public redo(): INotebookChange | undefined {
+		if (this._redoCells.length > 0) {
+			const cell = this._redoCells.pop();
+			if (cell) {
```
Same here - is this check actually necessary?

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/notebookHistory.ts:
```diff
@@ -0,0 +1,71 @@
+}
+
+export interface INotebookCellState {
+	action: INotebookChange,
```
Please add doc comments to all these interfaces. It's not really clear what "action" and "revert" are just from looking at this (why is revert part of a "cell state"? Should the interface maybe be something like `INotebookUndoRedoElement` similar to [IResourceUndoRedoElement](https://github.com/microsoft/azuredatastudio/blob/main/src/vs/platform/undoRedo/common/undoRedo.ts#L17)?)

Charles-Gagnon reviewed (changes requested):
A couple of comments, but I think the big thing here is that it seems like using the existing UndoRedoService would be a much better way of doing this so we can leverage all the stuff they've already written for that instead of re-implementing it ourselves.

barbaravaldez commented on src/sql/workbench/services/notebook/browser/models/notebookHistory.ts:
```diff
@@ -0,0 +1,71 @@
+
+import { ICellModel, MoveDirection } from 'sql/workbench/services/notebook/browser/models/modelInterfaces';
+
+export class NotebookHistory implements INotebookHistory {
```
I didn't know there was a  vscode UndoRedoService. I'll look into it

barbaravaldez commented on src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts:
```diff
@@ -279,13 +279,13 @@ export class TextCellComponent extends CellView implements OnInit, OnChanges {
 		let i = 0;
 		// walk up the node to the top and get it's index
-		n = n.previousSibling;
+		n = n?.previousSibling;
```
When I was testing the undo and redo I saw some console errors. I think we use this to get the cursor position. Let me check if I can repro

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/notebook.component.ts:
```diff
@@ -120,6 +120,23 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
+		// Check that there are no active elements to prevent the execution of undo/redo commands in other elements.
+		// We need to make sure that there are no active cells to prevent triggering undo/redo within code cells and check that text cells are not in edit mode.
+		if (!this.areElementsActive() && (this.activeCellId === '' || !this.cellToolbar.first.getEditCellAction().editMode)) {
+			if ((e.ctrlKey || e.metaKey) && e.shiftKey && e.key === 'z') {
```
ctrl+y is another commonly used redo shortcut (at least on Windows)

chlafreniere replied:
https://github.com/microsoft/azuredatastudio/blob/7e17bfb6ec543a2141112bcdc681ce49134ab2ea/src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts#L78

Let's be consistent with text cell components with key bindings.

(Can we store these conditions in a common location to be accessed both by notebook component and text cell component?)

barbaravaldez replied:
@chlafreniere Should we use notebookCommon.ts for that? 

https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/notebook/common/notebookCommon.ts

Charles-Gagnon commented:
Some issues I noticed while testing ad-hoc : 

1. The command palette commands to add a cell seem to bypass the logic here
2. It seems to work fine for the first untitled notebook I make, but then any others are ignored (the URI appears to be `Notebook-0` for all of them)
3. Undo after split cell just seems to be undoing the creation of the new cell

So make sure you're doing thorough testing of a bunch of different scenarios - it's fine to not have undo/redo for certain actions (like collapsing) for now, but ideally it should just no-op if something isn't supported.

barbaravaldez commented:
> Some issues I noticed while testing ad-hoc :
> 
> 1. The command palette commands to add a cell seem to bypass the logic here
> 2. It seems to work fine for the first untitled notebook I make, but then any others are ignored (the URI appears to be `Notebook-0` for all of them)
> 3. Undo after split cell just seems to be undoing the creation of the new cell
> 
> So make sure you're doing thorough testing of a bunch of different scenarios - it's fine to not have undo/redo for certain actions (like collapsing) for now, but ideally it should just no-op if something isn't supported.

@Charles-Gagnon, thank you so much for testing! I implemented the merge cells method to undo splitting a cell.
I'll be implementing the redo in a different PR because this scenario is more complex.
I fixed the other two issues that you found, as well.

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/cellEdit.ts:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	redo(): void {
+		// no op
```
Add explanation of why this is a no-op currently

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/cellEdit.ts:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	undo(): void {
+		this.model.mergeCells(this.firstCell, this.secondCell);
```
Make sure you're testing this for splitting in the middle of lines and such to ensure that we aren't adding/removing newline characters

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/notebookModel.ts:
```diff
@@ -664,7 +670,23 @@ export class NotebookModel extends Disposable implements INotebookModel {
+		let index = this._cells.findIndex(cell => cell.equals(cell));
+		if (index > -1) {
+			cell.source = cell.source.concat(...secondCell.source);
+			this._contentChangedEmitter.fire({
```
Do we not need to fire an event for deleting the other cell as well?

barbaravaldez replied:
The deleteCell method fires the event

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/notebookModel.ts:
```diff
@@ -674,9 +696,14 @@ export class NotebookModel extends Disposable implements INotebookModel {
+		if (addToUndoStack) {
+			this.undoService.pushElement(new AddCellEdit(this, cell, index));
+
+			cell.isEditMode = true;
```
Why is this logic in the addToUndoStack block only? This stuff shouldn't be tied to whether we add it to the history or not

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/notebookModel.ts:
```diff
@@ -817,6 +851,7 @@ export class NotebookModel extends Disposable implements INotebookModel {
 				let contents: nb.ICellContents = edit.cell as nb.ICellContents;
 				newCells.push(this._notebookOptions.factory.createCell(contents, { notebook: this, isTrusted: this._trustedMode }));
 			}
+			this.undoService.pushElement(new AddCellEdit(this, newCells[0], edit.range.start));
```
There may not be a new cell if edit.cell doesn't exist

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/notebook.component.ts:
```diff
@@ -118,6 +120,18 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
+		this._register(DOM.addDisposableListener(window, DOM.EventType.KEY_DOWN, (e: KeyboardEvent) => {
+			// Check that there are no active elements to prevent the execution of undo/redo commands in other elements.
+			// We need to make sure that there are no active cells to prevent triggering undo/redo within code cells and check that text cells are not in edit mode.
+			if (this.isActive() && !this.areElementsActive() && (this.activeCellId === '' || !this.cellToolbar.first.getEditCellAction().editMode)) {
```
Is this logic necessary because it's capturing the key events before they get handled by the editors? Because I would expect that when those handle the undo/redo that they stop propagation already.

Anyways this is probably fine to go in for now, but I want to follow up at looking at this more. We want to be careful with how we handle key events since historically we've had some major issues when we deviate from how VS Code handles these things. If possible registering this stuff as actual commands with keybindings would be very much preferred (since that way we actually let the user control the keybindings as well...)

If you could look into how Notebooks are hooking up their undo/redo that'd be a good place to start.

Charles-Gagnon replied:
@barbaravaldez Could you comment on my first question? What happens if you remove this?

barbaravaldez replied:
The only logic that it's strictly necessary is `this.isActive() && this.activeCellId === ''` to prevent the undo/redo from happening in other notebooks and to prevent the execution of undo/redo in the cell.

I'll remove the other conditions.

alanrenmsft commented on src/sql/workbench/contrib/notebook/browser/notebook.component.ts:
```diff
@@ -118,6 +120,18 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 		this._register(this._configurationService.onDidChangeConfiguration(e => {
 			this.doubleClickEditEnabled = this._configurationService.getValue('notebook.enableDoubleClickEdit');
 		}));
+		this._register(DOM.addDisposableListener(window, DOM.EventType.KEY_DOWN, (e: KeyboardEvent) => {
```
can we add the listener on the notebook cell element instead of window?

barbaravaldez replied:
@alanrenmsft I tried to add it to the notebook component but I cannot add listeners to divs. 
I'm not sure if we can add the listener to each notebook cell because we undo/redo when the notebook cell is not selected to prevent from triggering the undo/redo within the actual cell.

Charles-Gagnon replied:
Yeah - I think that this needs to go at the top level because we purposely run the logic only when nothing is active.

This goes back to my other comment though that generally I would expect keybindings to all be handled by the core command handling code instead of manually like this.

alanrenmsft commented on src/sql/workbench/contrib/notebook/browser/notebook.component.ts:
```diff
@@ -118,6 +120,18 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
+				if ((event.metaKey && event.shiftKey && event.keyCode === KeyCode.KEY_Z) || event.ctrlKey && event.keyCode === KeyCode.KEY_Y) {
+					this._model.redo();
+				} else if ((event.ctrlKey || event.metaKey) && event.keyCode === KeyCode.KEY_Z) {
+					this._model.undo();
```
you need to stop the event propagation after handling the keyboard events. right?

alanrenmsft commented on src/sql/workbench/services/notebook/browser/models/cellEdit.ts:
```diff
@@ -0,0 +1,77 @@
+
+export class AddCellEdit implements IResourceUndoRedoElement {
+	type: UndoRedoElementType.Resource = UndoRedoElementType.Resource;
+	label: string = 'Add Cell';
```
localize

alanrenmsft replied:
the comment also apply to other edit types

Charles-Gagnon commented on src/sql/workbench/contrib/notebook/browser/notebook.component.ts:
```diff
@@ -118,6 +120,20 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
+			if (this.isActive() && !this.areElementsActive() && (this.activeCellId === '' || !this.cellToolbar.first.getEditCellAction().editMode)) {
+				let event = new StandardKeyboardEvent(e);
+				if ((event.metaKey && event.shiftKey && event.keyCode === KeyCode.KEY_Z) || event.ctrlKey && event.keyCode === KeyCode.KEY_Y) {
+					event.stopPropagation();
```
You should use `EventHelper.stop(event, true)` to stop both propagation and cancel default

Charles-Gagnon commented on src/sql/workbench/services/notebook/browser/models/notebookModel.ts:
```diff
@@ -674,9 +696,13 @@ export class NotebookModel extends Disposable implements INotebookModel {
+		if (addToUndoStack) {
+			this.undoService.pushElement(new AddCellEdit(this, cell, index));
+			// Set newly created cell as active cell
+			this.updateActiveCell(cell);
```
This should be outside the if check as well, right?

barbaravaldez replied:
If I update the active cell outside of this condition, it will make the cell active. I'm not sure if this will be the best experience for undoing  multiple times, because the user would have to deselect the cell first and then try to undo again.

Charles-Gagnon replied:
Gotcha, yeah that makes sense. Could you add that to the comment?

Charles-Gagnon reviewed (approved):
A couple more things to address but otherwise looks good to get in
